Add base shared library functions

nikola-rados opened the pull request:
This PR introduces the first set of shared library functions.  The purpose of a shared library is to factor out common code from `Jenkinsfiles`.  In order to put this initial library together I factored out code from the [`crmprtd`](https://github.com/pacificclimate/crmprtd/blob/master/Jenkinsfile), [`climate-explorer-frontend`](https://github.com/pacificclimate/climate-explorer-frontend/blob/master/Jenkinsfile), [`climate-explorer-backend`](https://github.com/pacificclimate/climate-explorer-backend/blob/master/Jenkinsfile) and [`pdp`](https://github.com/pacificclimate/pdp/blob/master/Jenkinsfile) pipelines.  The overall changes are:
- Added 7 unique pipeline steps
- Added 5 utility classes
- Added 3 test classes (9 tests total) and a mock class

The factoring out greatly reduced the size of each individual `Jenkinsfile` (~70%) and makes them much more readable and understandable.  I'm hoping for lots of feedback! I want a good baseline for this library so that it is easy to work with and add to in the future.

## Notes for Reviewers
- Ignore any files that have to do with `gradle`, the places you will want to look at in `vars`, `src` and `test` (probably in that order).  The [`README`](https://github.com/pacificclimate/jenkins-library/tree/dev) covers what each of these directories are responsible for.
- The classes **must** implement `Serializable` to be used in a Jenkins pipeline, which is why you'll see it a lot
- There are a bunch of cases where I wanted to have optional arguments.  In order to do this in `groovy` I used `maps` + a utility function (`maps` are basically the groovy equivalent of python dictionaries).
- I'm looking to implement [`JenkinsPipelineUnit`](https://github.com/jenkinsci/JenkinsPipelineUnit) next.  It can be used to test the steps themselves rather than the classes the steps use.  For now I'm happy with some unit tests.
- I tested all of these steps on live pipelines and they worked!

nikola-rados committed:
Add title to README

nikola-rados committed:
Update README

nikola-rados committed:
Add demo step

nikola-rados committed:
Add docker build step

nikola-rados committed:
Add typing to declaration

nikola-rados committed:
Remove demo step

nikola-rados committed:
Add checkout step

nikola-rados committed:
Remove parameter for build function

nikola-rados committed:
Add default image name

nikola-rados committed:
Change default name to all lowercase

nikola-rados committed:
Add docker publishing steps

nikola-rados committed:
Add local image removal

nikola-rados committed:
Add docstrings

nikola-rados committed:
Add docker server uri default

nikola-rados committed:
Update docstring

nikola-rados committed:
Add parameter types where applicable

nikola-rados committed:
Rename variables for clarity

nikola-rados committed:
Fix up parameters

nikola-rados committed:
Fix spacing

nikola-rados committed:
Define return type

nikola-rados committed:
Remove credential id from params

nikola-rados committed:
Make docker registry a parameter

nikola-rados committed:
Add setup and supported steps to README

nikola-rados committed:
Update README

nikola-rados committed:
Add section on how to add new steps

nikola-rados committed:
Add more details

nikola-rados committed:
Rename steps

nikola-rados committed:
Remove default image name

nikola-rados committed:
Move credentials param

nikola-rados committed:
Add Python test step

nikola-rados committed:
Remove extra spacing

nikola-rados committed:
Add map for kwargs

nikola-rados committed:
Factor out python test suite code

nikola-rados committed:
Add push to pypi as step

nikola-rados committed:
Add utility package

nikola-rados committed:
Update to groovy styling

nikola-rados committed:
First overwritten variables

nikola-rados committed:
Add import

nikola-rados committed:
Move import to different spot

nikola-rados committed:
Update local import

nikola-rados committed:
Update utilities path

nikola-rados committed:
Update import name

nikola-rados committed:
Fix import issues

nikola-rados committed:
Remove required type

nikola-rados committed:
Change default to space

nikola-rados committed:
Change method to static

nikola-rados committed:
Add static to other util method

nikola-rados committed:
Add node test step

nikola-rados committed:
Add --pull for docker builds

nikola-rados committed:
Add docstrings to methods

nikola-rados committed:
Update README

nikola-rados committed:
Update docstrings

nikola-rados committed:
Move docker functionality to class

nikola-rados committed:
Add non cps tag

nikola-rados committed:
Add script element to class

nikola-rados committed:
Give class context variable

nikola-rados committed:
Apply non cps status again

nikola-rados committed:
Change context type

Commit:
Add tests section

Commit:
Add classes

Commit:
Remove incorrectly structured classes

Commit:
Update steps to use classes

Commit:
Add gradle artifacts

Commit:
Add tests for classes

Commit:
Fix typo

Commit:
Fix git command

Commit:
Fix shell param

Commit:
Update tag collection strategy

Commit:
Fix declaration order

Commit:
Add steps var to Utils class

Commit:
Pass pipeline context to Utils class

Commit:
Fix outdated method name

Commit:
Add new step

Commit:
Clean up step

Commit:
Fix bad var in step

Commit:
Fix typing mistake

Commit:
Fix typo

Commit:
Add option for non root image inspection

Commit:
Update pypi url

Commit:
Add case where -e install is available

Commit:
Add new param to step

Commit:
Test installation

Commit:
Testing for permission issue

Commit:
Revert egg changes

Commit:
Rename step for clarity

Commit:
Fix broken tests

Commit:
Update README.md

Commit:
Update docstrings

rod-glover commented on README.md:
```diff
@@ -1,2 +1,69 @@
+| Step | Description |
+|------|-------------|
+| [`buildDockerImage`](https://github.com/pacificclimate/jenkins-library/blob/dev/vars/buildDockerImage.groovy) | Build a Docker image onto a Docker server. |
+| [`codeCollection`](https://github.com/pacificclimate/jenkins-library/blob/dev/vars/codeCollection.groovy) | Uses the standard `checkout scm` command (only available for multi-branch pipelines) along with a git fetch. |
```
The names of all the other procedural steps here are imperative sentences, i.e., `<verb><object>`, which is an excellent and often recommended naming convention. So I suggest this be renamed `collectCode`.

rod-glover commented on README.md:
```diff
@@ -1,2 +1,69 @@
+| [`codeCollection`](https://github.com/pacificclimate/jenkins-library/blob/dev/vars/codeCollection.groovy) | Uses the standard `checkout scm` command (only available for multi-branch pipelines) along with a git fetch. |
+| [`publishDockerImage`](https://github.com/pacificclimate/jenkins-library/blob/dev/vars/publishDockerImage.groovy) | Publishes a docker image to a registry.  By default the name of the image should contain the registry to publish to in the form `your-registry/image-name:tag`.  There is an option to specify a registry with a URL. |
+| [`publishPythonPackage`](https://github.com/pacificclimate/jenkins-library/blob/dev/vars/publishPythonPackage.groovy) | Creates a python package out of the project and publishes it to pypi. |
+| [`pypiPublishable`](https://github.com/pacificclimate/jenkins-library/blob/dev/vars/pypiPublishable.groovy) | Checks branch conditions to see if the python package should be published. |  
```
Suggest name change to `isPypyPublishable`, following a common convention for boolean tests.

jameshiebert replied:
I think you mean `isPypiPublishable`.

rod-glover commented on README.md:
```diff
@@ -1,2 +1,69 @@
+This section outlines how the project is strucuted. If you are looking to add functionality please read this section before coding.  
+
+### Steps
+All of the steps are located in the [`vars/`](https://github.com/pacificclimate/jenkins-library/tree/dev/vars) directory.  Each `.groovy` file corresponds to a step where the step name is the name of the file.  This is the name you would use in a `Jenkinsfile` to use the step.  In the file there **must** be a `call` function:
```
Is this directory name (and location) a Jenkins convention? It seems odd and non-intuitive relative to the other code directory (i.e., `src/org/pcic`).

nikola-rados replied:
I agree with your comment but it is the convention.  `src`, `test` and `vars` are meant to be top level directories.  As for the `org/pcic` path, I've seen some differences in my exploration of other libraries.  Sometimes people use `src/com/yourcompany`, others use `src/org/yourcompany` and I've even seen deeper paths like `src/com/groovy/yourcompany`.  We _could_ break with convention and maybe group the classes into more descriptive package names like `src/classes` and `src/util`.

rod-glover replied:
Until I know why the convention exists, I'd hesitate to break it. Although the convention looks a little odd and heavy-handed to me, it might be best to keep it if it is widely followed.

rod-glover commented on vars/buildDockerImage.groovy:
```diff
@@ -0,0 +1,25 @@
+         image = docker.build(imageName, processed.buildArgs)
+     }
+
+     return image
```
Can you just put the return inside the `withDockerServer` block?

rod-glover commented on vars/codeCollection.groovy:
```diff
@@ -0,0 +1,11 @@
+import org.pcic.GitUtils
```
Change (file) name to `collectCode`?

rod-glover commented on vars/publishDockerImage.groovy:
```diff
@@ -0,0 +1,34 @@
+            }
+        }
+    }
+}
```
Nice clear code!

rod-glover commented on vars/publishPythonPackage.groovy:
```diff
@@ -0,0 +1,39 @@
+        pytainer.inside {
+            pytils.preparePackage(pip)
+
+            withCredentials([usernamePassword(credentialsId: credentialsId, usernameVariable: 'USERNAME', passwordVariable: 'PASSWORD')]) {
```
This long line would be easier to read with some newlines, if that is syntactically possible.

rod-glover commented on vars/pypiPublishable.groovy:
```diff
@@ -0,0 +1,18 @@
+        return true
+    } else {
+        return false
+    }
```
Suggest replacing if statement with `return (BRANCH_NAME == 'master' && !tags.isEmpty())`

rod-glover commented on vars/pypiPublishable.groovy:
```diff
@@ -0,0 +1,18 @@
+import org.pcic.GitUtils
```
Rename file to `isPypiPublishable`?

rod-glover commented on src/org/pcic/DockerUtils.groovy:
```diff
@@ -0,0 +1,41 @@
+    ArrayList determineTags(String branchName, ArrayList gitTags) {
+        ArrayList publishTags = []
+
+        if(!gitTags.isEmpty() && branchName == 'master') {
```
This is the condition in `pypiPublishable`. Should that function be used here so that any change to that condition is automatically used here?

rod-glover commented on src/org/pcic/DockerUtils.groovy:
```diff
@@ -0,0 +1,41 @@
+                         'pdp': '-u root --volumes-from pdp_data --env-file /storage/data/projects/comp_support/jenkins/pdp_envs/pdp_deployment.env']
+        String args
+        if(available.containsKey(project)) {
+            args = available[project]
```
Just `return available[project]` here.

rod-glover commented on src/org/pcic/GitUtils.groovy:
```diff
@@ -0,0 +1,27 @@
+            retTags = []
+        }
+
+        return retTags
```
I don't want to step all over your code style here, but in general the early-return pattern for this kind of computation is easier to read. Any code that requires you to build up a mental model of the contents of a variable that then has to be carried forward to where it is acted on (in this case, returned) adds mental overhead.

Instead you can do a pattern like this, where what is returned is directly stated. It's simpler to understand, and shorter.

```
if (gitTags.isEmpty()) {
    return []
}

return gitTags.trim().split('\n')
```

rod-glover commented on src/org/pcic/PythonUtils.groovy:
```diff
@@ -0,0 +1,51 @@
+            pip = 'pip'
+        }
+
+        return pip
```
Use early-return pattern

rod-glover commented on src/org/pcic/util/Utils.groovy:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    Map processParams(Map expected, Map args) {
+        def unknownKeys = (expected.keySet() + args.keySet()) - expected.keySet()
```
If `+` is set union, and `-` is set difference, then this expression is equivalent to `args.keySet() - expected.keySet()`. Is that what you meant?

nikola-rados replied:
Good catch, that's not supposed to be there.

rod-glover commented on src/org/pcic/util/Utils.groovy:
```diff
@@ -0,0 +1,35 @@
+            name = steps.env.BRANCH_NAME
+        }
+
+        return name
```
Use early-return pattern

rod-glover commented on src/org/pcic/util/Utils.groovy:
```diff
@@ -0,0 +1,35 @@
+            throw new Exception("Key(s) ${unknownKeys} not available in ${expected.keySet()}")
+        }
+
+        return expected + args
```
Doing a little judicious guessing here (instead of looking up the language ref), it looks like what this function does is update `expected` with `args`, and throw an error if `args` contains any key not in `expected`. Yes? In any case, a comment explaining what it computes would be helpful. And maybe a more explanatory name than `processParams` would be helpful; something like `argsWithDefaults` perhaps.

rod-glover commented on test/org/pcic/DockerUtilsTest.groovy:
```diff
@@ -0,0 +1,101 @@
+
+        // assert
+        assert result == expected_tags
+        assert result.size() == expected_size
```
Checking the size is redundant here, I think.

rod-glover commented on test/org/pcic/DockerUtilsTest.groovy:
```diff
@@ -0,0 +1,101 @@
+
+        // assert
+        assert result == expected_tags
+        assert result.size() == expected_size
```
Checking the size is redundant here, I think.

rod-glover commented on test/org/pcic/DockerUtilsTest.groovy:
```diff
@@ -0,0 +1,101 @@
+
+        // assert
+        assert result == expected_tags
+        assert result.size() == expected_size
```
Checking the size is redundant here, I think.

rod-glover commented on test/org/pcic/DockerUtilsTest.groovy:
```diff
@@ -0,0 +1,101 @@
+        // assert
+        assert result == expected_tags
+        assert result.size() == expected_size
+    }
```
Is the commenting pattern `// prepare` ... `// execute` ... `// assert` typical or idiomatic in Jenkins/groovy? The comments seem unnecessary; the whitespace already separates the three blocks adequately, and also `assert` calls out the test expectations proper. I see this throughout.

rod-glover commented on test/org/pcic/DockerUtilsTest.groovy:
```diff
@@ -0,0 +1,101 @@
+        String project = 'badKey'
+
+        // execute
+        String result = dockerUtils.getContainerArgs(project)
```
Is there a reason these 5 lines are better than this 1?

```
String result = dockerUtils.getContainerArgs('badKey')
```

or even, if it is allowable for a function

```
dockerUtils.getContainerArgs('badKey')
```

rod-glover replied:
I get that there is a test code pattern being followed here, but I am not convinced it is delivering value for bulk. In fact, particularly in simple instances like this one, I'd say it makes the code noticeably less easily grasped.

rod-glover commented on test/org/pcic/GitUtilsTest.groovy:
```diff
@@ -0,0 +1,31 @@
+
+        // assert
+        assert result == expected_tags
+        assert result.size() == expected_size
```
Redundant

rod-glover commented on test/org/pcic/util/MockSteps.groovy:
```diff
@@ -0,0 +1,15 @@
+            retVal = "1.0.0\nsome-tag"
+        }
+
+        return retVal
```
Does this return `undefined` or some such if the if-condition is false?

rod-glover reviewed (changes requested):
Overall, this is great. The separation of concerns is very clean and the code, some style issues notwithstanding, is clear and simple to understand.

I do have some feedback on code style; see comments. 

Generally speaking, there are many places where the reader has to remember the value of a variable to figure out what is returned, when instead you could just return the value immediately. Often this can be accomplished in the early-return pattern, which turns a cascade of if-then-else into a series of `if (cond) { return value }`, followed by a final `return value` for the default (final else) case.

There is a version of this in the tests, too, where you set up both arguments and expected return value for the function under test as variables, and then use those in the invocation and very straightforward test assertions. Unless there is something complicated about computing the args or expected values, it is usually much clearer what the test is doing if these are all (or mostly) inlined.

Commit:
Rename steps

Commit:
Early return

Commit:
Shrink line length

Commit:
Factor out common condition into method

Commit:
Change logic to early-return

Commit:
Add default param handler to Utils

Commit:
Add test for Utils class

Commit:
Update method names and add docstrings

Commit:
Add test for exception case

Commit:
Update args processing

Commit:
Remove size tests

Commit:
Simplify test code

Commit:
Remove comments in tests

Commit:
Add jf

Commit:
Clean up mock method

Commit:
Use gradle plugin to run tests

nikola-rados commented:
Thank you for all the feedback @rod-glover!  I've addressed all the comments you made.  The most notable change is the update to how each of the `steps` handle default variables (you can find it in `Utils.groovy`).  I'm happy to have a discussion if you believe there is a better way to handle getting and updating default args.  I've also added a short `Jenkinsfile` that runs the tests.

Let me know if you see anything else that should be changed!

Commit:
Test push for webhook

rod-glover commented on src/org/pcic/util/Utils.groovy:
```diff
@@ -0,0 +1,70 @@
+     */
+    Map getUpdatedArgs(ArrayList keys, Map argsToUpdate) {
+        return updateArgs(getDefaultArgs(keys), argsToUpdate)
+    }
```
Nice

rod-glover commented on src/org/pcic/util/Utils.groovy:
```diff
@@ -0,0 +1,70 @@
+                        pypiUrl: 'https://pypi.pacificclimate.org/',
+                        pythonVersion: 3]
+        return defaults.subMap(keys)
+    }
```
Nice. So `subMap` exists, thank goodness!

rod-glover commented on src/org/pcic/util/Utils.groovy:
```diff
@@ -0,0 +1,70 @@
+        // `defaults` and `updates` will be updated to have the value from
+        // `updates`.  Other key, value pairs are left as is.
+        return defaults + updates
+    }
```
I like how you have partitioned out the updater separately.

rod-glover commented on src/org/pcic/util/Utils.groovy:
```diff
@@ -0,0 +1,70 @@
+    boolean isPublishable(String branchName, ArrayList tags) {
+        return (branchName == 'master' && !tags.isEmpty())
+    }
+}
```
Utils is looking good!

rod-glover commented on test/org/pcic/util/MockSteps.groovy:
```diff
@@ -0,0 +1,20 @@
+        // Return some tags
+        if(args.script == 'git tag --contains' && args.containsKey('returnStdout')) {
+            return "1.0.0\nsome-tag"
+        }
```
What does it (or should it) return if the condition is false? What happens in this language when you don't have an explicit `return` statement?

nikola-rados replied:
That particular condition is when the `returnStdout` flag is set to `true`.  In that case some string is expected to be returned.  Otherwise the `sh` command doesn't return anything.  In groovy if have something like this:
```groovy
def myMethod(val) {
    if (val > 1) {
        return 'some-string'
    }
}

def test = myMethod(0)
```
The class of `test` will be a `NullObject`.  You can still call `myMethod()` without assigning the output a  to a variable and it won't complain either way (return value, or no return value).

Edit: the method name

rod-glover commented on vars/buildDockerImage.groovy:
```diff
@@ -0,0 +1,21 @@
+ */
+ def call(String imageName, Map options = [:]) {
+     Utils utils = new Utils(this)
+     Map args = utils.getUpdatedArgs([serverUri, buildArgs], options)
```
This is exactly the pattern I was suggesting. Still looks good to me.

rod-glover commented on src/org/pcic/util/Utils.groovy:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /**
+     * Returns a submap given a list of keys
```
Suggest rewrite:

Returns a map of default arguments, given a list of keys. Default arguments are common to many steps and classes, so are defined once here. Because only a subset of the arguments is required or valid for any given step, the keys of those arguments are passed in and only those defaults are returned.

rod-glover replied:
See note above regarding purpose vs. method.

rod-glover commented on src/org/pcic/util/Utils.groovy:
```diff
@@ -0,0 +1,70 @@
+
+    /**
+     * This method takes in a list of keys and a map, and returns a map with
+     * updates values.
```
Comments should state the purpose and result (why and what) much more than the method (how). This reads more like method than purpose.

rod-glover commented on test/org/pcic/DockerUtilsTest.groovy:
```diff
@@ -0,0 +1,76 @@
+        ArrayList result = dockerUtils.determineTags(branchName, gitTags)
+
+        assert result == expected_tags
+    }
```
Yeah, this is just as clear and 1/2 the length. Win!

rod-glover commented:
Oops, didn't add an overall comment to the 2nd review. Which is:

Good work, @nikola-rados . I'd suggest only a couple of tweaks to your comments, which could convey more useful information than they do.

Apart from that, LGTM!

Commit:
Fix comments

Commit:
Update comments again

Commit:
Fix default lists

Commit:
Factor out docker image name construction

Commit:
Update return type

Commit:
Fix string output

Commit:
Add tests for Utils

Commit:
Add tests for DockerUtils

nikola-rados commented:
@rod-glover I updated the comments.  I also did some live testing and found a couple mistakes, those were also addressed.  I added some more unit tests where applicable.  Let me know if you find anything that needs to be changed, otherwise I'm ready to merge!

rod-glover commented:
Sorry, meant to approve the changes in the last round. Merge at will!
